Implement Insert into Kafka

elonazoulay opened the pull request:
Convert key columns to KafkaColumnHandle

<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Rewrite of #7407

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues

Rewriting commit by commit from the previous attempt.
Once the approach looks good we can reopen and overwrite the original PR or convert this one to a non-draft PR.

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(X) Release notes are required, with the following suggested text:
Add the syntax for specifying the key subject, value subject and key columns:
key-subject, value-subject: when a SubjectNameStrategy other than TopicName is used
key-columns: specify key column name and type when schema registry is not used for the key, ex. a simple long or string key type (this is a common use case).

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

elonazoulay commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaTableHandle.java:
```diff
@@ -54,6 +54,7 @@
     private final Optional<String> messageDataSchemaLocation;
     private final Optional<String> keySubject;
     private final Optional<String> messageSubject;
+    private final Optional<List<KafkaColumnHandle>> keyColumns;
```
Previously this was a string which encoded <column name>:<column type> and was decoded multiple times which is unnecessary.

elonazoulay replied:
Changed this to `keyColumn` as the usage is to have a message subject stored in schema registry but a raw key, like a long or varchar id. The only way to specify the type (that I am aware of) is to encode it in the table name:
ex. 
```
SELECT col FROM kafka_table&key-column=my_id:varchar
```
Also added this to the exception message if it's incorrectly specified.

vlad-lyutenko commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/ConfluentModule.java:
```diff
@@ -146,13 +149,13 @@ public void configure(Binder binder)
-            });
-            encoderFactoriesByName.addBinding(ProtobufRowEncoder.NAME).toInstance((session, dataSchema, columnHandles) -> {
+            encoderFactoriesByName.addBinding(AvroRowEncoder.NAME).to(ConfluentRowEncoderFactory.class).in(SINGLETON);
+            encoderFactoriesByName.addBinding(KafkaRowDecoderFactory.NAME).to(KafkaSerializerRowEncoder.Factory.class).in(SINGLETON);
```
I am really new to this topic, just started review, but cannot get why we bind decoder factory to encoder factory.

elonazoulay replied:
Good catch, thanks!

elonazoulay commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/RowDecoderFactory.java:
```diff
@@ -13,10 +13,12 @@
 public interface RowDecoderFactory
 {
-    RowDecoder create(Map<String, String> decoderParams, Set<DecoderColumnHandle> columns);
+    RowDecoder create(Map<String, String> decoderParams, Set<DecoderColumnHandle> columns, ConnectorSession session);
```
TODO: extract to a separate commit if we are ok with this approach:
This is to support a common use case where an avro record has no fields.
The new behavior of avro 1.11.1 no longer allows missing fields so it has to be explicitly set by another decoder (EmptyFieldHandlingAvroDecoder).

vlad-lyutenko commented:
lgtm

hashhar commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/DispatchingRowDecoderFactory.java:
```diff
@@ -32,9 +33,9 @@ public DispatchingRowDecoderFactory(Map<String, RowDecoderFactory> factories)
     }
 
-    public RowDecoder create(String dataFormat, Map<String, String> decoderParams, Set<DecoderColumnHandle> columns)
+    public RowDecoder create(String dataFormat, Map<String, String> decoderParams, Set<DecoderColumnHandle> columns, ConnectorSession session)
```
can you extract the param addition to some preparatory commit?

hashhar commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/file/FileWriterSchemaReader.java:
```diff
@@ -0,0 +1,54 @@
+import static io.trino.plugin.kafka.KafkaErrorCode.KAFKA_SCHEMA_ERROR;
+import static java.lang.String.format;
+
+public class FileWriterSchemaReader
```
nit: I'd probably rename `ContentSchemaReader` to `ContentSchemaProvider`

Then you can call this `FileWriteSchemaProvider` and rename the existing read path one to `FileReadSchemaProvider`.

`FileWriter` led to some confusion initially - like we write back the content schema to disk.

hashhar commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/AvroSchemaConverter.java:
```diff
@@ -208,7 +210,7 @@ private Optional<Type> convertRecord(Schema schema)
                     return Optional.empty();
                 case ADD_DUMMY:
-                    return Optional.of(RowType.from(ImmutableList.of(new RowType.Field(Optional.of(DUMMY_FIELD_NAME), BooleanType.BOOLEAN))));
+                    return Optional.of(DUMMY_ROW_TYPE);
```
so it's now a RowType column called `dummy` with a single boolean field called `$dummy`?

i.e. `dummy ROW(dummy BOOLEAN)`?

elonazoulay replied:
Yes: if there was an empty avro subrecord called `foo` then the schema will be `foo ROW("$dummy" BOOLEAN)`. I thought it would be more clear to prefix the dummy field with `$`. wdyt?

hashhar commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaRecordSet.java:
```diff
@@ -171,7 +171,6 @@ private boolean nextRow(ConsumerRecord<byte[], byte[]> message)
             long timeStamp = message.timestamp() * MICROSECONDS_PER_MILLISECOND;
 
             Optional<Map<DecoderColumnHandle, FieldValueProvider>> decodedKey = keyDecoder.decodeRow(keyData);
-            // tombstone message has null value body
```
why remove the comment?

elonazoulay replied:
added it back, thanks!

hashhar commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaPageSink.java:
```diff
@@ -126,7 +126,13 @@ public CompletableFuture<?> appendPage(Page page)
 
-            expectedWrittenBytes += keyBytes.length + messageBytes.length;
+            /**
+             * KafkeProducerBatch returns -1 for bytes length if data is null
```
```suggestion
             * KafkaProducerBatch returns -1 for bytes length if data is null
```

hashhar replied:
BTW this seems like an existing bugfix - what prevents the connector from writing empty messages today?

elonazoulay replied:
The encoders used in this pr will return null, could be that the encoders already in use will just return 0, I will verify.

elonazoulay replied:
This is an existing bugfix. If an empty message is written then bytes length will return 0, if it is null then it will return -1. This prevents the error from the check where the bytes written matches expected written bytes in the `finish` method.

hashhar commented on plugin/trino-kafka/pom.xml:
```diff
@@ -98,11 +98,21 @@
         </dependency>
 
+        <dependency>
+            <groupId>io.confluent</groupId>
+            <artifactId>kafka-avro-serializer</artifactId>
+        </dependency>
+
         <dependency>
             <groupId>io.confluent</groupId>
             <artifactId>kafka-schema-registry-client</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>io.confluent</groupId>
+            <artifactId>kafka-schema-serializer</artifactId>
```
cannot be used in compile scope due to license issues

hashhar replied:
nevermind, those two modules are Apache licensed. It's so confusing with Confluent stuff. 😭

hashhar reviewed (commented):
skimmed at the moment, overall shape looks good, some questions mainly about the dependency issue

Will do another deeper pass in a while

(some comments might not make too much sense since I did not review commit by commit yet)

hashhar commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/decoder/EmptyFieldHandlingAvroRowDecoder.java:
```diff
@@ -0,0 +1,98 @@
+                        return entry;
+                    }
+                })
+                .collect(toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue)));
```
some reason to not use Guava map?

elonazoulay replied:
changed everything to guava.

elonazoulay replied:
yes, also commented below (for lists): maps may contain non-null key with null values. Also added a comment in the code to explain.

hashhar commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/decoder/EmptyFieldHandlingAvroRowDecoder.java:
```diff
@@ -0,0 +1,98 @@
+                .collect(toUnmodifiableMap(Map.Entry::getKey, Map.Entry::getValue)));
+    }
+
+    private boolean isDummyRowType(Type type)
```
this assumes no user would have a `ROW("$dummy" BOOLEAN)` as well - maybe add a comment.

Q: There can be 0 or more than 1 empty fields which we encode like this?

hashhar replied:
Also does the description for the session property need to be updated since it says the field is named `dummy` while it'll be `$dummy` now.

elonazoulay replied:
There can be multiple subfields of a record that can be encoded like this, i.e. if the actual schema contains multiple empty subrecords. For a record it is only considered a "dummy" row if it has only 1 boolean field named "$dummy". Did that answer your question?

hashhar commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/TrinoToAvroSchemaConverter.java:
```diff
@@ -0,0 +1,74 @@
+            return Schema.create(Schema.Type.BYTES);
+        }
+        else if (trinoType instanceof DateType) {
+            return Schema.create(Schema.Type.INT);
```
where are the logical type annotations added? i.e. how would readers of the schema know it's an Avro logical `Date` and not a plain INT.

hashhar replied:
Same for the following two branches.

elonazoulay replied:
Thanks! Added a commit for this (`Add support for time and timestamp avro logical types`) and the tests for keys are in `TestKeyColumnsConnectorSmokeTest.testSupportedKeyColumnTypes` and messages in `TestKafkaWithConfluentLatestConnectorSmokeTest.testTemporalTypes`.

hashhar commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/AbstractConfluentRowEncoder.java:
```diff
@@ -0,0 +1,246 @@
+        Optional<Schema> baseSchema = schema.getTypes().stream()
+                .filter(fieldSchema -> fieldSchema.getType() != Schema.Type.NULL)
+                .map(nonNullSchema -> {
+                    if (nonNullSchema.isUnion()) {
```
what if it's a union type of `string` and `Foo`? it seems this would arbitrarily return `string` or `Foo`.

Since we don't support such unions we should disallow it here.

elonazoulay replied:
Good point! Now it will only support UNION(NULL, <TYPE>) where TYPE can also be UNION(NULL, <TYPE>) but there must only be one type.

hashhar commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/AbstractConfluentRowEncoder.java:
```diff
@@ -0,0 +1,246 @@
+            for (int i = 0; i < arrayBlock.getPositionCount(); i++) {
+                list.add(getValue(elementType, arrayBlock, i, elementAvroType));
+            }
+            return Collections.unmodifiableList(list);
```
specific reason to avoid Guava collections in this class?

elonazoulay replied:
Switching to guava:) Saw the suggestion to use guava since if we use it everywhere then we avoid extra copies and also preserve order for builders, etc.

elonazoulay replied:
Update: had to switch back because lists and maps may contain null values, for maps it may be a non-null key with a null value.

hashhar commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/AbstractConfluentRowEncoder.java:
```diff
@@ -0,0 +1,246 @@
+    @Override
+    public void close()
+    {
+        // Clear thread local encoder
```
did you mean to add some code here?

elonazoulay replied:
Removed - that was left over from an older version

hashhar reviewed (commented):
another pass, yet to read table description supplier and new smoke test in detail

kokosing commented:
LGTM for below commits. Can you please extract them to separate PR and we can merge them fast?
<img width="451" alt="image" src="https://github.com/trinodb/trino/assets/7569403/4861e521-f278-49a0-b326-64fc28411a52">

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/decoder/KafkaRowDecoderFactory.java:
```diff
@@ -0,0 +1,118 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class KafkaRowDecoderFactory
```
Can you please provide tests for that class? It looks unused in the commit which introduces it.

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/TrinoToAvroSchemaConverter.java:
```diff
@@ -0,0 +1,124 @@
+        else if (isTimeMicrosType(trinoType)) {
+            return timeMicros().addToSchema(Schema.create(Schema.Type.LONG));
+        }
+        else {
```
nit: no need for `else`

kokosing commented on docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -475,7 +475,7 @@ schema registry. You must also configure the additional properties in the follow
         * ``IGNORE`` - Ignore structs with no fields. This propagates to parents.
```
This commit can also be extracted and merged fast.

kokosing commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/DispatchingRowDecoderFactory.java:
```diff
@@ -15,6 +15,7 @@
 
```
Extract this commit too :)

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/decoder/EmptyFieldHandlingAvroRowDecoder.java:
```diff
@@ -0,0 +1,99 @@
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static java.util.Objects.requireNonNull;
+
+public class EmptyFieldHandlingAvroRowDecoder
```
tests?

elonazoulay replied:
There are 2 tests in `TestKafkaWithConfluentLatestConnectorSmokeTest`:
`testEmptyStructWithMarkStrategy` and `testEmptyStructWithIgnoreStrategy`

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaPageSink.java:
```diff
@@ -126,7 +126,13 @@ public CompletableFuture<?> appendPage(Page page)
+            /**
+             * KafkaProducerBatch returns -1 for bytes length if data is null
+             * @see org.apache.kafka.clients.producer.internals.ProducerBatch#tryAppend
+             */
```
Was this causing some user visible defect? Can you have a test for that? Is it possible that `finish()` method would fail because of incorrect value in `expectedWrittenBytes`, how can you trigger that to test your fix?

elonazoulay replied:
Will update: this did cause test failures w empty field handling, will call it out explicitly, thanks!

elonazoulay replied:
The init methods fail for BaseConnectorTest and BaseConnectorSmokeTest. Here is the error without this change:
```
[ERROR] Tests run: 724, Failures: 1, Errors: 0, Skipped: 425, Time elapsed: 122.4 s <<< FAILURE! -- in TestSuite
[ERROR] io.trino.plugin.kafka.TestKafkaWithConfluentLatestConnectorSmokeTest.init -- Time elapsed: 13.61 s <<< FAILURE!
io.trino.testing.QueryFailedException: Cannot read the array length because "keyBytes" is null
```

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaSplitManager.java:
```diff
@@ -15,7 +15,7 @@
 
```
Extract this too :)

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaPageSinkProvider.java:
```diff
@@ -79,13 +79,17 @@ public ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHa
                 session,
```
Extract this too :)

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/AbstractConfluentRowEncoder.java:
```diff
@@ -0,0 +1,246 @@
+            return type.getSlice(block, position).toByteBuffer();
+        }
+
+        if (type instanceof ArrayType) {
```
else if?

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/AbstractConfluentRowEncoder.java:
```diff
@@ -0,0 +1,246 @@
+            }
+            return recordBuilder.build();
+        }
+        else {
```
else not needed

kokosing commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaWithConfluentLatestConnectorSmokeTest.java:
```diff
@@ -0,0 +1,1299 @@
+    {
+        assertThatThrownBy(super::testInsert)
+                .hasMessage("Cannot test INSERT without CREATE TABLE, the test needs to be implemented in a connector-specific way");
+        // TODO implement the test for Kafka
```
reminder

elonazoulay replied:
Custom tests for insert are in `TestKafkaWithConfluentLatestConnectorSmokeTest`.

kokosing commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaWithConfluentLatestConnectorSmokeTest.java:
```diff
@@ -0,0 +1,1299 @@
+import static org.testng.Assert.assertTrue;
+
+public class TestKafkaWithConfluentLatestConnectorSmokeTest
+        extends BaseConnectorSmokeTest
```
can you use `BaseConnectorTest`?

kokosing commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaWithConfluentLatestConnectorSmokeTest.java:
```diff
@@ -0,0 +1,1299 @@
+        throw new SkipException("TODO");
+    }
+
+    private static ImmutableMap.Builder<String, String> schemaRegistryAwareProducer(TestingKafka testingKafka)
```
inline

elonazoulay replied:
This method is used many times in the BaseConnectorTest and BaseConnectorSmokeTest - extracted to a base class - lmk if you think it's better to inline it, I can make that change quickly.

kokosing commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaWithConfluentLatestConnectorSmokeTest.java:
```diff
@@ -0,0 +1,1299 @@
+                .put(SCHEMA_REGISTRY_URL_CONFIG, testingKafka.getSchemaRegistryConnectString());
+    }
+
+    private static String toDoubleQuoted(String tableName)
```
move utilities to the bottom of the class

kokosing commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaWithConfluentLatestConnectorSmokeTest.java:
```diff
@@ -0,0 +1,1299 @@
+    private void waitUntilTableExists(Session session, String tableName)
+    {
+        Failsafe.with(
+                        RetryPolicy.builder()
```
extract variable for policy

kokosing commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaWithConfluentLatestConnectorSmokeTest.java:
```diff
@@ -0,0 +1,1299 @@
+    }
+
+    @Test
+    public void testSupportedKeyColumnTypes()
```
Can we move such low level tests out of the SmokeTest? To separate test class?

elonazoulay replied:
Extracted to `TestKeyColumnsConnectorSmokeTest`.

kokosing reviewed (commented):
Initial skim. Let's iterate and merge what we can and I will review it again.

wendigo commented:
@elonazoulay do you plan to rebase this PR?

elonazoulay commented:
> @elonazoulay do you plan to rebase this PR?

Hi, apologies for the delay, yes. I am testing the changes in #18980 and if tests pass I will update this branch. Wanted to make sure I addressed the review comments, etc.

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/KafkaMetadata.java:
```diff
@@ -100,6 +100,7 @@ public KafkaTableHandle getTableHandle(ConnectorSession session, SchemaTableName
                         kafkaTopicDescription.getMessage().flatMap(KafkaTopicFieldGroup::getDataSchema),
```
Should we add docs changes for this commit?

elonazoulay replied:
Yep! Added, lmk if you have any suggestions.

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/decoder/EmptyFieldHandlingAvroRowDecoder.java:
```diff
@@ -0,0 +1,99 @@
+            blockBuilder.buildEntry(fieldBuilders -> {
+                fieldBuilders.get(0).appendNull();
+            });
+            return blockBuilder.build().getObject(0, Block.class);
```
Can we have a static block that we would always return? Do we need to create it each time?

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/decoder/EmptyFieldHandlingAvroRowDecoder.java:
```diff
@@ -0,0 +1,99 @@
+{
+    public static final String NAME = "avro";
+
+    private static final FieldValueProvider NULL_FIELD_VALUE_PROVIDER = new FieldValueProvider()
```
named static inner class somewhere below?

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/decoder/EmptyFieldHandlingAvroRowDecoder.java:
```diff
@@ -0,0 +1,99 @@
+                    if (isDummyRowType(entry.getKey().getType())) {
+                        return new AbstractMap.SimpleImmutableEntry<>(entry.getKey(), NULL_FIELD_VALUE_PROVIDER);
+                    }
+                    else {
```
else is not needed

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/decoder/EmptyFieldHandlingAvroRowDecoder.java:
```diff
@@ -0,0 +1,99 @@
+    public Optional<Map<DecoderColumnHandle, FieldValueProvider>> decodeRow(byte[] data)
+    {
+        return delegate.decodeRow(data).map(decodedRow -> decodedRow.entrySet().stream()
+                .map(entry -> {
```
can we avoid nested lambdas?

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/decoder/EmptyFieldHandlingAvroRowDecoder.java:
```diff
@@ -0,0 +1,99 @@
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue)));
+    }
+
+    // An empty field encoded as a dummy row has exactly one field named "$dummy" with boolean type.
```
is this comment still accurate? It was renamed recently

elonazoulay replied:
Thanks! Updated the field name, lmk if that works.

kokosing commented on plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/decoder/EmptyFieldHandlingAvroRowDecoder.java:
```diff
@@ -0,0 +1,99 @@
+            return false;
+        }
+        RowType.Field field = rowType.getFields().get(0);
+        return field.getType() == BOOLEAN && field.getName().map(name -> name.equals(DUMMY_FIELD_NAME)).orElse(false);
```
DUMMY_FIELD_NAME should be renamed to EMPTY_FIELD_MARKER in preparatory commit

kokosing commented on lib/trino-record-decoder/src/main/java/io/trino/decoder/avro/AvroColumnDecoder.java:
```diff
@@ -200,6 +214,24 @@ public boolean getBoolean()
         @Override
         public long getLong()
         {
+            if (columnType instanceof TimeType) {
```
tests

kokosing commented on plugin/trino-kafka/pom.xml:
```diff
@@ -73,11 +73,21 @@
 
+        <dependency>
+            <groupId>io.confluent</groupId>
+            <artifactId>kafka-avro-serializer</artifactId>
```
What is the license of this?

elonazoulay replied:
It is an Apache 2.0 license.

kokosing commented on plugin/trino-kafka/pom.xml:
```diff
@@ -73,11 +73,21 @@
 
+        <dependency>
+            <groupId>io.confluent</groupId>
+            <artifactId>kafka-schema-serializer</artifactId>
```
What is the license of this?

kokosing commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaWithConfluentLatestConnectorTest.java:
```diff
@@ -0,0 +1,139 @@
+        return KafkaWithConfluentSchemaRegistryQueryRunner.builder(testingKafka)
+                .setTables(ImmutableList.of(ORDERS, LINE_ITEM, NATION, REGION, PART, PART_SUPPLIER, SUPPLIER, CUSTOMER))
+                .setExtraKafkaProperties(ImmutableMap.<String, String>builder()
+                        .put("kafka.confluent-subjects-cache-refresh-interval", "1ms")
```
can this test be executed in parallel?

elonazoulay replied:
I will try. If I see failures in ci then I may have to break it up though. Will update. lmk if that's what you meant.

kokosing commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/TestKafkaWithConfluentLatestConnectorTest.java:
```diff
@@ -0,0 +1,139 @@
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestKafkaWithConfluentLatestConnectorTest
```
I think you can remove `Latest`

kokosing commented on plugin/trino-kafka/src/test/java/io/trino/plugin/kafka/BaseKafkaWithConfluentLatestSmokeTest.java:
```diff
@@ -0,0 +1,163 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.testng.Assert.assertTrue;
+
+public abstract class BaseKafkaWithConfluentLatestSmokeTest
```
We use smoke tests if we want to test different deployment (different setup with kafka). Here you are using the same kafka and Trino as in `TestKafkaWithConfluentLatestConnectorSmokeTest`, so I would move all tests to of this class and tests from all implementation of this class to `TestKafkaWithConfluentLatestConnectorTest`

mosabua commented:
I assume that you are still working on this @elonazoulay

vytautas-karpavicius commented:
Are there any plans on merging this? This very desired functionality when building kafka centric ingestion pipelines.

We have tested this PR on our internal fork and it works great!

The only issue I have noticed - it can not to be configured with `kafka.hide-internal-columns=false`. Otherwise it tries to insert into internal columns. But it is easy to workaround, by configuring another catalog without this property.

mosabua commented:
I added the stale-ignore label and we will continue to watch manually.
